astropy/io/misc/pyarrow/csv.py

```diff
@@ -0,0 +1,402 @@
+        if input_file_stripped is not None:
+            input_file = input_file_stripped
+
+    table_pa = csv.read_csv(
+        input_file,
+        parse_options=parse_options,
+        read_options=read_options,
+        convert_options=convert_options,
+    )
+    table_apt = convert_pa_table_to_astropy_table(table_pa)
+
+    return table_apt
+
+
+def get_pyarrow_csv():
+    """Helper function to import pyarrow and pyarrow.csv."""
+    if not HAS_PYARROW:
+        raise ModuleNotFoundError(
+            "pyarrow is required to read and write pyarrow.csv files"
+        )
+    import pyarrow as pa
+    from pyarrow import csv
+
+    return pa, csv
+
+
+def convert_pa_array_to_numpy(
+    arr: Union["pa.Array", "pa.ChunkedArray"],
+) -> "npt.NDArray":
+    """
+    Convert a PyArrow array to a NumPy array.
+
+    Parameters
+    ----------
+    arr : pyarrow.Array, pyarrow.ChunkedArray
+        The PyArrow array to be converted.
+
+    Returns
+    -------
+    np.ndarray or np.ma.MaskedArray
+        The converted NumPy array. If the input array contains null values, a masked
+        array is returned with nulls masked out.
+
+    Notes
+    -----
+    - If the input array is of string type, it delegates the conversion to
+      `convert_pa_string_array_to_numpy`.
+    - If the input array contains null values, they are replaced with 0 before
+      conversion, and a masked array is returned with the null positions masked.
+    - If the input array does not contain null values, the result is an ndarray view of
+      the underlying data.
+    """
+    import pyarrow.compute as pc
+
+    is_string = arr.type == "string"
+
+    if arr.null_count == 0:
+        # No nulls, just return an ndarray view of the pyarray
+        if is_string:
+            max_length = pc.max(pc.utf8_length(arr))
+            out = np.empty(len(arr), dtype=f"U{max_length}")
+            out[:] = arr.to_numpy()
+        else:
+            out = arr.to_numpy()
+    else:
+        # Return a Masked Array copy, with nulls filled with zero.
+        data = arr.fill_null("" if is_string else 0).to_numpy()
+        if is_string:
+            data = data.astype(str)
+        # pa Bool array may not be zero-copyable, so set zero_copy_only=False to avoid
+        # an exception.
+        mask = arr.is_null().to_numpy(zero_copy_only=False)
+        out = np.ma.array(data, mask=mask, copy=False)
+
+    return out
+
+
+def convert_pa_table_to_astropy_table(table_pa: "pa.Table") -> "Table":
+    """
+    Convert a PyArrow Table to an Astropy Table.
+
+    Parameters
+    ----------
+    table_pa : pa.Table
+        The PyArrow Table to be converted.
+
+    Returns
+    -------
+    Table
+        Converted astropy Table.
+    """
+    from astropy.table import Table
+
+    columns = {
+        name: convert_pa_array_to_numpy(col)
+        for name, col in zip(table_pa.column_names, table_pa.itercolumns())
+    }
+    out = Table(columns, copy=False)
```

Comment:
I'd again tend to just return directly here, but obviously no big deal.

Reply:
I use this pattern a lot because it facilitates debugging. From the context of this function I can easily examine `out` along with the other variables defined in the function.

Reply:
OK, that is a nice reason. We better make sure we never enable the ruff rule that would remove it!